retryrpc/api.go

```diff
@@ -0,0 +1,134 @@
+// Package retryrpc provides a client and server RPC model which survives
+// lost connections on either the client or the server.
+package retryrpc
+
+import (
+	"container/list"
+	"sync"
+	"time"
+)
+
+// Request is the structure sent
+type Request struct {
+	Len       uint64 // Length of this struct including len field
+	RequestID uint64 // Unique ID of this request
+	MountID   uint64 // Mount ID
+	JReq      []byte // JSON containing request
+}
+
+// Response is the structure returned
+type Response struct {
+	Len uint64 // Length of this struct including len field
+	Res []byte // JSON response - only valid if err != nil
+	Err error  // Status of request
+}
+
+type completedRequest struct {
+	completedTime time.Time
+	request       *Request
+}
+
+// Server tracks the state of the server
+type Server struct {
+	sync.Mutex
+	completedTTL      time.Duration // How long a completed request stays on queue
+	serviceMap        map[string]*func(request interface{}, response interface{})
+	connections       *list.List                   // List of outstanding connections
+	pendingRequests   map[string]*Request          // Request which have not been completed yet.
+	completedRequests map[string]*completedRequest // Request which have been completed.  Will age out
+	ipaddr            string                       // IP address server listens too
+	port              int                          // Port of server
+}
+
+// NewServer creates the Server object
+func NewServer(ttl time.Duration, ipaddr string, port int) *Server {
+	s := &Server{}
+	s.serviceMap = make(map[string]*func(request interface{}, response interface{}))
+	s.connections = list.New()
+	s.completedTTL = ttl
+	s.ipaddr = ipaddr
+	s.port = port
+	return s
+}
+
+// Register creates the map of server methods
+func (server *Server) Register() (err error) {
+
+	// TODO - build the serviceMap similiar to how net/rpc Register does it...
+	// serviceMap looks like serviceMap["RpcPing"]*rpcPing()
+	return
+}
+
+// Run server loop, accept connections, read request, run RPC method and
+// return the results.
+func (server *Server) Run() {
+	// TODO - Algorithm - Standard server stuff
+	// 1. goroutine which accepts new connection and creates goroutine W
+	// 2. goroutine W adds request to server.pendingRequest, unmarshals
+	//    request, processes RPC, puts request on server.completedRequest
+	//    and under same lock removes from server.pendingRequest.
+	// 3. Results are sent back to client if socket still up
+	// 4. request stays on server.completedRequest until s.completedTTL exceeded
+
+}
+
+// Close stops the server
+func (server *Server) Close() {
+	// TODO - blocks until the server is shut down
+
+}
+
+// Client methods
+
+// Client tracking structure
```

Comment:
So I gather this `retryrpc` package will be used by both package jrpcfs (or its replacement/peer) and program pfsagentd, correct? Sounds great!

Reply:
yes - that is correct